### DISCUSSION TITLE
Very slow progress of optimization code

### DISCUSSION THREAD

**Discussion by NaumanHaider** [yalmip/YALMIP#930] [Problems using YALMIP]
[Troubleshoot file.zip](https://github.com/yalmip/YALMIP/files/6697657/Troubleshoot.file.zip)

Hej,

I am trying to simulate the attached code but the performance is very slow when integrated with simulink. I have checked with profiler as well and it shows that 98% of the time is spent executing optimizer.m function. I am unsure how to increase the execution speed here. The speed becomes even (100 times) slower  as I am supposed to integrate this simulink model with AMESIM (Siemens Software) for my thesis work. 
Also I am restricted to using Matlab interpreted function block due to dynamic nature of my plant model. Yalmip doesn't allow the use of standard matlab function block as it needs to know the size of problem beforehand. Is there any way the constraints can be initialized to desired size lmi instead of as an empty array prior to execution with values updated during execution. This I think might speed up execution as standard function block is comparatively faster than interpreted one.

It will be highly appreciable if you could suggest any improvements or point out to what is it that I can look into to improve execution speed.

Thanks and Regards,

Nauman Khurshid Haider

**Comment by johanlofberg** [comment]
You sent the cache (slc) instead of the model (slx)

**Reply by NaumanHaider** [reply]
[Troubleshoot file.zip](https://github.com/yalmip/YALMIP/files/6701043/Troubleshoot.file.zip)
Sorry for the mistake, I have added the correct file to the folder now. Thanks

**Reply by johanlofberg** [reply]
You are setting up a new optimizer object in every sample, which makes no sense and sort of defeats the whole purpose of using an optimizer object.

The actuall call to the optimizer, which is what you have to do in the samples after the first one, is very fast

```
K>> tic;controller{x};toc
Elapsed time is 0.021811 seconds
```

Everything explained here
https://yalmip.github.io/example/simulink/

**Reply by NaumanHaider** [reply]
The problem is in my case both the objective function (the penalty matrix R) and constraints (due to system matrices B and D) are changing for each sample time due to changing value of states and IL. I can't use the same controller for all time steps.

**Reply by johanlofberg** [reply]
Nothing prevents you from having other stuff than the state as parameters. However having R directly as a parameter might lead to an effiicient model to work with in compiled stage as it will be cubic

The best model is probably to add new a variable e, use e'*e in the objective, add e == diag(r)*u where r is a new parameter (which will be the square-roots of the wanted penalty matrix elements)

Same thing with B. Simply parameterize the changing elements in that matrix and have those as parameters too. Also done in the last MPC example  https://yalmip.github.io/example/standardmpc/

**Reply by NaumanHaider** [reply]
I just ran the code after make the changes suggested by you, now the run time is a lot faster (1000 sec takes 1600 seconds now) compared previously. Thanks a lot for your advice.
